prepare.sh

```diff
@@ -1,7 +1,64 @@
-#!/bin/sh
+#!/bin/bash
 
 set -x
 
 yum -y install git podman
 
-git clone --single-branch --branch=master https://github.com/ceph/ceph-csi /opt/build
+ARGUMENT_LIST=(
+    "ref"
+    "workdir"
+    "gitrepo"
+)
+
+opts=$(getopt \
+    --longoptions "$(printf "%s:," "${ARGUMENT_LIST[@]}") $(printf "help::")" \
+    --name "$(basename "${0}")" \
+    --options "" \
+    -- "$@"
+)
+
+eval set --$opts
+
+while true; do
+    case "${1}" in
+    --help)  
+        shift
+        echo "Options:"
+        echo "--help|-h                 specify the flags"
+        echo "--ref                     specify the reference of pr"
+        echo "--workdir                 specify the working directory"
+        echo "--gitrepo                 specify the git repository" 
+        echo " "
+        echo "Sample Usage:"
+        echo "./prepare.sh --gitrepo=https://github.com/example --workdir=/opt/build --ref=pull/123/head"
+        ;;
+    --gitrepo)  
+        shift
+        gitrepo=${1}
+        ;;
+    --workdir)  
+        shift
+        workdir=${1}
+        ;;
+    --ref)  
+        shift
+        ref=${1} 
+        #extracting Pull Id from passed reference
+        ghprbPullId="$(echo ${ref} | cut -d '/' -f2)"         
```

Comment:
Why discard a standard ref and force it into a PR style?

Reply:
My apologies, I'm new with scripting and couldn't completely get your point. Could you please elaborate?

Reply:
Nothing too specific to scripting here. :-)
The command line option is --ref, which implies that any ref that is valid to git could be passed in and used. However, this line of code discards everything but the PR number (assuming  the github style of pr refs). Instead of silently discarding information I recommend either renaming the option, to say "--prnum" and explicitly making it depend on github style PRs or making the code general enough to check out any generic ref. Since the code for pr handling is already done the former is certainly going to be easier than the latter. I see that @nixpanic asked for an option of this naming scheme but I want to make sure what's (implicitly) advertised by the option's name matches what is implemented by the code.

Reply:
OK, now I went back an re-read everything.... and its sort of going to work but in an odd fashion.

Anything that is passed as a ref goes to the "git fetch origin ..." command but the actual local  name that is created has tried to get a pr number out of it and then names it "pr_BLAH".
If my cli is called with --ref=master you end up executing `git fetch origin master:pr_master`. If my cli was --ref=foo/bar `git fetch origin foo/bar:pr_bar. I just find this confusing.

Reply:
I might be wrong here, but this script is meant to be triggered only during creation of a new Pull Request on the ceph-csi repo by Jenkins. Given the use case, I expected the references to be only of type "pull/ghprbPullId/head". Should we expect a different format of reference in our use case? Let me know what you think, if you recommend, I'll make the changes :)

Reply:
The `--ref=...` will mostly be used for passing the reference to the Pull Request. However in future we will likely have nightly (or periodic) jobs that run from the master branch or others. These would be jobs that combine ceph-csi with other projects, think of updated versions of side-cars, an unreleased version of Kubernetes or very time-consuming tests that do not make sense to be part of the requirements for every PR (certain workload/functionality tests as [CDI](https://github.com/kubevirt/containerized-data-importer)).

So, try not to use `ghprbPullId` in this script, but only use `ref` or similar. In case `--ref` is not passed, you can set it to `master` as default.